pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -449,31 +668,88 @@ func (p *testPlanner) changeVersionSteps(
 // hooks while the cluster version is changing. At the end of this
 // process, we wait for all the migrations to finish running.
 func (p *testPlanner) finalizeUpgradeSteps(
-	fromVersion, toVersion *clusterupgrade.Version, scheduleHooks bool,
+	fromVersion, toVersion *clusterupgrade.Version, scheduleHooks, virtualClusterSetup bool,
 ) []testStep {
-	p.currentContext.System.Finalizing = true
-	p.currentContext.System.Stage = RunningUpgradeMigrationsStage
-	runMigrations := p.newSingleStep(allowUpgradeStep{})
-	var mixedVersionStepsDuringMigrations []testStep
+	p.currentContext.SetFinalizing(true)
+	p.currentContext.SetStage(RunningUpgradeMigrationsStage)
+	runSystemMigrations := p.newSingleStep(allowUpgradeStep{
+		virtualClusterName: install.SystemInterfaceName,
+	})
+
+	steps := []testStep{runSystemMigrations}
+
+	var mixedVersionStepsDuringTenantMigrations []testStep
 	if scheduleHooks {
-		mixedVersionStepsDuringMigrations = p.hooks.MixedVersionSteps(p.currentContext, p.prng, p.isLocal)
+		mixedVersionSteps := p.hooks.MixedVersionSteps(p.currentContext, p.prng, p.isLocal)
+
+		switch p.deploymentMode {
+		case SystemOnlyDeployment:
+			steps = append(steps, mixedVersionSteps...)
+
+		case SharedProcessDeployment:
+			// If we are in a multi-tenant deployment, we run some
+			// mixed-version hooks while the system is inalizing, and the
+			// remaining hooks while the tenant is finalizing.
+			if len(mixedVersionSteps) > 0 {
+				idx := p.prng.Intn(len(mixedVersionSteps))
+
+				mixedVersionStepsDuringSystemMigrations := mixedVersionSteps[:idx]
+				steps = append(steps, mixedVersionStepsDuringSystemMigrations...)
+
+				mixedVersionStepsDuringTenantMigrations = mixedVersionSteps[idx:]
+			}
+		}
 	}
 
-	waitForMigrations := p.newSingleStep(
+	steps = append(steps, p.newSingleStep(
 		waitForStableClusterVersionStep{
-			nodes:              p.crdbNodes,
+			nodes:              p.currentContext.System.Descriptor.Nodes,
 			timeout:            p.options.upgradeTimeout,
 			desiredVersion:     versionToClusterVersion(toVersion),
 			virtualClusterName: install.SystemInterfaceName,
 		},
-	)
+	))
+
+	// At this point, we just upgraded the storage cluster; we might
+	// need to patch up the `system.tenant_settings` table in order for
+	// the tenant upgrade to work.
+	steps = append(steps, p.maybeDeleteAllTenantsVersionOverride(toVersion)...)
+
+	switch p.deploymentMode {
+	case SharedProcessDeployment:
+		if virtualClusterSetup {
+			steps = append(
+				steps,
+				p.newSingleStep(allowUpgradeStep{
```

Comment:
It's currently enforced in the upgrade machinery that tenants are not allowed to upgrade until the system has finished upgrading. We could change the timing of the `preserve_downgrade_option` on the tenant. In that case, we would be verifying that tenants know not to start (auto-)upgrading because the system is not upgraded yet.

I think this should be done in future work, perhaps the existing `PreserveDowngradeOptionRandomizer` could be adapted to change tenant steps too.